lib/ruby_lsp/setup_bundler.rb

```diff
@@ -186,9 +186,22 @@ def run_bundle_install(bundle_gemfile = @gemfile)
         env["BUNDLE_PATH"] = File.expand_path(env["BUNDLE_PATH"], @project_path)
       end
 
+      path_parts = if Gem.win_platform?
+        ENV["Path"] || ENV["PATH"] || ENV["path"] || ""
+      else
+        ENV["PATH"] || ""
+      end.split(File::PATH_SEPARATOR)
+
+      bin_directory = File.expand_path("bin", @project_path)
+      bundle_binstub = File.join(@project_path, "bin", "bundle")
+
       # If there's a Bundler version locked, then we need to use that one to run bundle commands, so that the composed
       # lockfile is also locked to the same version. This avoids Bundler restarts on version mismatches
-      base_bundle = if @bundler_version
+      base_bundle = if File.exist?(bundle_binstub) && path_parts.any? do |path|
+        File.expand_path(path, @project_path) == bin_directory
+      end
+        bundle_binstub
+      elsif @bundler_version
```

Comment:
It might just be me not being used to inline conditional assignment, but I found this a little hard to read.

What do you think of this:

```ruby
base_bundle = setup_bundle

def setup_bundle
  if bundle_binstub_exists_and_matches?
    bundle_binstub
  elsif @bundler_version
    env["BUNDLER_VERSION"] = @bundler_version.to_s
    install_bundler_if_needed
    "bundle _#{@bundler_version}_"
  else
    "bundle"
  end
end

def bundle_binstub_exists_and_matches?
  File.exist?(bundle_binstub) && path_parts.any? do |path|
    File.expand_path(path, @project_path) == bin_directory
  end
end
```

I am also curious, other than personal preference, are there any tangible drawbacks to splitting it into multiple methods?

Reply:
I extracted the `base_bundle_command` as a method. Regarding drawbacks, it will mostly depend on the performance requirements of the code you are writing.

Dispatching a method in Ruby is always more expensive than not doing it. So, for example, if you have a high iteration loop that's critical for performance, it might be a good move to reduce the number of method dispatches that happen inside to maximize performance.

In this particular case, where the method is invoked once for the whole life cycle of the language server, it makes absolutely no difference. And it does improve readability, thanks for pointing that out!